File: lib/app/services/trading/trade_order_test.exs
defmodule TradeOrderTest do
  use DataCase

  test "create/1" do
    # given
    trading_account_id = TestUtils.acc()

    # when given a new limit order
    MatchingService.create(trading_account_id, "BTC_EUR", :LIMIT, :SELL, 20.10, 10, :GTC)

    # then should be saved
    assert MatchingServiceTestHelpers.get_sell_book_order_count() === 1

    # when given a new market order
    MatchingService.create(trading_account_id, "BTC_EUR", :MARKET, :SELL, 10, :GTC)

    # then should be saved
    assert MatchingServiceTestHelpers.get_sell_book_order_count() === 2

    # when given a stop loss order
    MatchingService.create(trading_account_id, "BTC_EUR", :STOPLOSS, :SELL, 20.10, 10, :GTC)

    # then should be not be saved to order book
    assert MatchingServiceTestHelpers.get_sell_book_order_count() === 2

    # when given a stop limit order
    MatchingService.create(trading_account_id, "BTC_EUR", :STOPLIMIT, :SELL, 20.10, 10, :GTC)

    # then should be not be saved to order book
    assert MatchingServiceTestHelpers.get_sell_book_order_count() === 2
  end
end

File: pkg/services/matching/trade_order_test.go
package matching

func (assert *ServiceTestSuite) TestCreateTradeOrderBook() {
	// given

	// when given a new limit order
	ProcessTradeOrder(assert.instrumentAccount1,
		"BTC_EUR",
		"LIMIT", Sell,
		20.10,
		10, "GTC")

	// then should be saved
	assert.Equal(1, GetSellBookOrderCount())

	// when given a new market order
	ProcessTradeOrder(assert.instrumentAccount1,
		"BTC_EUR", "MARKET", Sell, 0, 10, "GTC")

	// then should be saved
	assert.Equal(2, GetSellBookOrderCount())

	// when given a stop loss order
	ProcessTradeOrder(assert.instrumentAccount1,
		"BTC_EUR", "STOPLOSS", Sell, 20.10, 10, "GTC")

	// then should be not be saved to order book
	assert.Equal(2, GetSellBookOrderCount())

	// when given a stop limit order
	ProcessTradeOrder(assert.instrumentAccount1, "BTC_EUR", "STOPLIMIT", Sell, 20.10, 10, "GTC")

	// then should be not be saved to order book
	assert.Equal(2, GetSellBookOrderCount())
}
